### PR TITLE
Fix KVM not detecting containerd preload

### PR DESCRIPTION
**Pull request by spowelljr**
### Cause
We were doing a `sudo systemctl restart containerd` right before we checked if the preloaded images are present. It was common on KVM that containerd wouldn't finish restarting before running `sudo crictl images --output json` which would fail, and would return `false` for the images being preloaded, resulting in the images be redownloaded unnecessarily.

### Fix
Added a retry to the `sudo crictl images --output json` command so if containerd is still restarting we don't report that the preloaded images aren't present.

#### First time start
Before: 59s
After: 38.6s
**20.4s speedup (35%)** 

#### Normal start
Before: 48.7s
After: 30.5s
**18.2s speedup (37%)**

Also fixes the `TestErrorSpam/setup` test failure on KVM containerd due to the `❌  Unable to load cached images` that was being output before.

**Before:**
```
$ time minikube start --driver kvm --container-runtime containerd
😄  minikube v1.32.0 on Ubuntu 20.04
✨  Using the kvm2 driver based on user configuration
💾  Downloading driver docker-machine-driver-kvm2:
    > docker-machine-driver-kvm2-...:  65 B / 65 B [---------] 100.00% ? p/s 0s
    > docker-machine-driver-kvm2-...:  13.01 MiB / 13.01 MiB [ 100.00% ? p/s 0s
💿  Downloading VM boot image ...
    > minikube-v1.32.1-1700142131...:  65 B / 65 B [---------] 100.00% ? p/s 0s
    > minikube-v1.32.1-1700142131...:  293.18 MiB / 293.18 MiB  100.00% 240.78 
👍  Starting control plane node minikube in cluster minikube
💾  Downloading Kubernetes v1.28.4 preload ...
    > preloaded-images-k8s-v18-v1...:  436.27 MiB / 436.27 MiB  100.00% 235.22 
🔥  Creating kvm2 VM (CPUs=2, Memory=6000MB, Disk=20000MB) ...
📦  Preparing Kubernetes v1.28.4 on containerd 1.7.9 ...
❌  Unable to load cached images: loading cached images: stat /home/powellsteven/.minikube/cache/images/amd64/registry.k8s.io/kube-proxy_v1.28.4: no such file or directory
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔗  Configuring bridge CNI (Container Networking Interface) ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

real	0m59.048s
user	0m4.741s
sys	0m2.471s
```

**After:**
```
$ time minikube start --driver kvm --container-runtime containerd
😄  minikube v1.32.0 on Ubuntu 20.04
✨  Using the kvm2 driver based on user configuration
💾  Downloading driver docker-machine-driver-kvm2:
    > docker-machine-driver-kvm2-...:  65 B / 65 B [---------] 100.00% ? p/s 0s
    > docker-machine-driver-kvm2-...:  13.01 MiB / 13.01 MiB [ 100.00% ? p/s 0s
💿  Downloading VM boot image ...
    > minikube-v1.32.1-1700142131...:  65 B / 65 B [---------] 100.00% ? p/s 0s
    > minikube-v1.32.1-1700142131...:  293.18 MiB / 293.18 MiB  100.00% 283.11 
👍  Starting control plane node minikube in cluster minikube
💾  Downloading Kubernetes v1.28.4 preload ...
    > preloaded-images-k8s-v18-v1...:  436.27 MiB / 436.27 MiB  100.00% 212.95 
🔥  Creating kvm2 VM (CPUs=2, Memory=6000MB, Disk=20000MB) ...
📦  Preparing Kubernetes v1.28.4 on containerd 1.7.9 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔗  Configuring bridge CNI (Container Networking Interface) ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

real	0m38.648s
user	0m4.669s
sys	0m2.661s
```

### CODE REVIEW

File: pkg/minikube/cruntime/containerd.go [hunk 1]
```diff
@@ -568,13 +569,20 @@ func (r *Containerd) Restart() error {
 
 // containerdImagesPreloaded returns true if all images have been preloaded
 func containerdImagesPreloaded(runner command.Runner, images []string) bool {
-	rr, err := runner.RunCmd(exec.Command("sudo", "crictl", "images", "--output", "json"))
-	if err != nil {
+	var rr *command.RunResult
+
+	imageList := func() (err error) {
```
Comment (medyagh) [inline pkg/minikube/cruntime/containerd.go:574]: hm... I am trying to think, is there a better condition to wait for instead of retrying?

**Comment by prezha** [comment]
hey @spowelljr i think the actual problem is in (how we handle/expect) the `ctr` output:

from logs:
```
I1126 22:25:11.544084   26459 cache_images.go:88] LoadImages start: [registry.k8s.io/kube-apiserver:v1.28.4 registry.k8s.io/kube-controller-manager:v1.28.4 registry.k8s.io/kube-scheduler:v1.28.4 registry.k8s.io/kube-proxy:v1.28.4 registry.k8s.io/pause:3.9 registry.k8s.io/etcd:3.5.9-0 registry.k8s.io/coredns/coredns:v1.10.1 gcr.io/k8s-minikube/storage-provisioner:v5]
...
I1126 22:25:11.544147   26459 image.go:134] retrieving image: registry.k8s.io/kube-proxy:v1.28.4
...
I1126 22:25:11.547965   26459 image.go:177] daemon lookup for registry.k8s.io/kube-proxy:v1.28.4: Error response from daemon: No such image: registry.k8s.io/kube-proxy:v1.28.4
...
I1126 22:25:11.963349   26459 ssh_runner.go:195] Run: /bin/bash -c "sudo ctr -n=k8s.io images check | grep registry.k8s.io/kube-proxy:v1.28.4"
...
I1126 22:25:12.937500   26459 cache_images.go:116] "registry.k8s.io/kube-proxy:v1.28.4" needs transfer: "registry.k8s.io/kube-proxy:v1.28.4" does not exist at hash "83f6cc407eed88d214aad97f3539bde5c8e485ff14424cd021a3a2899304398e" in container runtime
...
I1126 22:25:12.937543   26459 cri.go:218] Removing image: registry.k8s.io/kube-proxy:v1.28.4
...
I1126 22:25:13.509165   26459 ssh_runner.go:195] Run: sudo /usr/bin/crictl rmi registry.k8s.io/kube-proxy:v1.28.4
...
I1126 22:25:14.276116   26459 cache_images.go:286] Loading image from: /home/prezha/.minikube/cache/images/amd64/registry.k8s.io/kube-proxy_v1.28.4
...
W1126 22:25:14.361961   26459 out.go:239] ❌  Unable to load cached images: loading cached images: stat /home/prezha/.minikube/cache/images/amd64/registry.k8s.io/kube-proxy_v1.28.4: no such file or directory
❌  Unable to load cached images: loading cached images: stat /home/prezha/.minikube/cache/images/amd64/registry.k8s.io/kube-proxy_v1.28.4: no such file or directory
...
```
so, the above underlying error:
>I1126 22:25:12.937500   26459 cache_images.go:116] "registry.k8s.io/kube-proxy:v1.28.4" needs transfer: "registry.k8s.io/kube-proxy:v1.28.4" does not exist at hash "83f6cc407eed88d214aad97f3539bde5c8e485ff14424cd021a3a2899304398e" in container runtime

comes from [this](https://github.com/kubernetes/minikube/blob/b1c5d95745e0c01b04ab55c5bf7f874332a219f3/pkg/minikube/machine/cache_images.go#L117) line, and that effectively come from this [block](https://github.com/kubernetes/minikube/blob/b1c5d95745e0c01b04ab55c5bf7f874332a219f3/pkg/minikube/cruntime/containerd.go#L251-L258)

looking at the crt output, we have:
```
$ sudo ctr -n=k8s.io images list
REF                                                                                                             TYPE                                                      DIGEST                                                                  SIZE      PLATFORMS                                                                     LABELS
...
registry.k8s.io/kube-proxy:v1.28.4                                                                              application/vnd.docker.distribution.manifest.list.v2+json sha256:e63408a0f5068a7e9d4b34fd72b4a2b0e5512509b53cd2123a37fc991b0ef532 23.4 MiB  linux/amd64,linux/arm64,linux/ppc64le,linux/s390x                             io.cri-containerd.image=managed
...
registry.k8s.io/kube-proxy@sha256:e63408a0f5068a7e9d4b34fd72b4a2b0e5512509b53cd2123a37fc991b0ef532              application/vnd.docker.distribution.manifest.list.v2+json sha256:e63408a0f5068a7e9d4b34fd72b4a2b0e5512509b53cd2123a37fc991b0ef532 23.4 MiB  linux/amd64,linux/arm64,linux/ppc64le,linux/s390x                             io.cri-containerd.image=managed
...
sha256:83f6cc407eed88d214aad97f3539bde5c8e485ff14424cd021a3a2899304398e                                         application/vnd.docker.distribution.manifest.list.v2+json sha256:e63408a0f5068a7e9d4b34fd72b4a2b0e5512509b53cd2123a37fc991b0ef532 23.4 MiB  linux/amd64,linux/arm64,linux/ppc64le,linux/s390x                             io.cri-containerd.image=managed
...
```
which has image `ref` and image `digest` sha in each line, whereas we want image ref and image `id` sha on the same like, and hence that fails (as there's image id on a separate "line")

to prove the point, `cricrl` gives both image ref and id on the same line (from the same minikube cluster):
```
$ sudo crictl images --no-trunc --digests
IMAGE                                     TAG                  DIGEST                                                                    IMAGE ID                                                                  SIZE
...
registry.k8s.io/kube-proxy                v1.28.4              sha256:e63408a0f5068a7e9d4b34fd72b4a2b0e5512509b53cd2123a37fc991b0ef532   sha256:83f6cc407eed88d214aad97f3539bde5c8e485ff14424cd021a3a2899304398e   24.6MB
...
```

i've tested the following:
```
diff --git a/pkg/minikube/cruntime/containerd.go b/pkg/minikube/cruntime/containerd.go
index aa99dc696..6001a33a0 100644
--- a/pkg/minikube/cruntime/containerd.go
+++ b/pkg/minikube/cruntime/containerd.go
@@ -248,12 +248,11 @@ func (r *Containerd) Disable() error {

 // ImageExists checks if image exists based on image name and optionally image sha
 func (r *Containerd) ImageExists(name string, sha string) bool {
-       c := exec.Command("/bin/bash", "-c", fmt.Sprintf("sudo ctr -n=k8s.io images check | grep %s", name))
+       c := exec.Command("sudo", "ctr", "-n=k8s.io", "images", "check")
        rr, err := r.Runner.RunCmd(c)
-       if err != nil {
-               return false
-       }
-       if sha != "" && !strings.Contains(rr.Output(), sha) {
+       if err != nil ||
+               !strings.Contains(rr.Output(), name) ||
+               (sha != "" && !strings.Contains(rr.Output(), sha)) {
                return false
        }
        return true
```
and it worked:
```
I1127 23:06:49.993578  115508 cache_images.go:88] LoadImages start: [registry.k8s.io/kube-apiserver:v1.28.4 registry.k8s.io/kube-controller-manager:v1.28.4 registry.k8s.io/kube-scheduler:v1.28.4 registry.k8s.io/kube-proxy:v1.28.4 registry.k8s.io/pause:3.9 registry.k8s.io/etcd:3.5.9-0 registry.k8s.io/coredns/coredns:v1.10.1 gcr.io/k8s-minikube/storage-provisioner:v5]
...
I1127 23:06:49.993720  115508 image.go:134] retrieving image: registry.k8s.io/kube-proxy:v1.28.4
...
I1127 23:06:49.994852  115508 image.go:177] daemon lookup for registry.k8s.io/kube-proxy:v1.28.4: Error response from daemon: No such image: registry.k8s.io/kube-proxy:v1.28.4
...
I1127 23:06:50.401707  115508 ssh_runner.go:195] Run: sudo ctr -n=k8s.io images check
...
I1127 23:06:50.906378  115508 cache_images.go:123] Successfully loaded all cached images
...
```

before:
```
real    0m58.370s
user    0m1.498s
sys     0m0.620s
```

after:
```
real    0m33.686s
user    0m1.356s
sys     0m0.404s
```

i'm happy to make a separate pr to fully test this, if it would help

File: pkg/minikube/cruntime/containerd.go (hunk 1 above)
Comment (prezha) [reply pkg/minikube/cruntime/containerd.go:574]: we probably don't have to wait - pls see the comment [below](https://github.com/kubernetes/minikube/pull/17658#issuecomment-1828807988)

**Comment by spowelljr** [comment]
@prezha I wonder if you're running into another issue, if I output the error in the file `pkg/minikube/cruntime/containerd.go` line 571 I get the following

```
sudo crictl images --output json: Process exited with status 1
stdout:

stderr:
time="2023-11-27T23:43:08Z" level=fatal msg="validate service connection: validate CRI v1 image API for endpoint \"unix:///run/containerd/containerd.sock\": rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/containerd/containerd.sock: connect: no such file or directory\""
```

I believe what's happening is that the `sudo crictl images --output json` command I'm referening is failing because containerd is not running due to the restart. And if my command fails it, it marks preloads as not loaded, and then it ends up hitting the code you're referencing which actually outputs the error message we're seeing. So it looks like we have two issues, the containerd not being ready due to restart, and then the logic you're referencing is also broken. Me fixing the containerd restart is causing the code you're referencing to not at all, but it's still broken.

**Comment by prezha** [comment]
@spowelljr  i think you're right: we probably have two separate issues here - the one that you observed happens before this other code that i referred to is hit
so it makes sense to wait for the containerd to come up after the restart, and if we wait here, then i think it's reasonable to retry

**Pull request merged**